packages/shared/src/hooks/feed/useAutorotatingAds.ts

```diff
@@ -0,0 +1,113 @@
+import { useCallback, useContext, useEffect, useMemo } from 'react';
+import {
+  useQuery,
+  useQueryClient,
+  type InfiniteData,
+  focusManager,
+} from '@tanstack/react-query';
+import {
+  useInView,
+  type InViewHookResponse,
+} from 'react-intersection-observer';
+import { useLogContext } from '../../contexts/LogContext';
+import { ActiveFeedContext } from '../../contexts';
+import { useFeature } from '../../components/GrowthBookProvider';
+import { featureAutorotateAds } from '../../lib/featureManagement';
+import type { Ad } from '../../graphql/posts';
+import { apiUrl } from '../../lib/config';
+import { generateAdLogEventKey } from './useLogImpression';
+import { disabledRefetch } from '../../lib/func';
+import { RequestKey } from '../../lib/query';
+
+export type InViewRef = InViewHookResponse['ref'];
+
+export const useAutorotatingAds = (
+  ad: Ad,
+  index: number,
+  feedIndex: number,
+  ref: InViewRef,
+): {
+  ref: InViewRef;
+} => {
+  const autorotateAds = useFeature(featureAutorotateAds);
+  const { logEventEnd } = useLogContext();
+  const { queryKey: feedQueryKey } = useContext(ActiveFeedContext);
+  const queryClient = useQueryClient();
+  const { ref: inViewRef, inView } = useInView({
+    threshold: 0.5,
+    skip: !autorotateAds,
+  });
+
+  const refs = useCallback(
+    (node: HTMLElement) => {
+      ref(node);
+      inViewRef(node);
+    },
+    [inViewRef, ref],
+  );
+
+  const queryKey = useMemo(
+    () => [RequestKey.Ads, ...feedQueryKey],
+    [feedQueryKey],
+  );
+
+  const fetchNewAd = useCallback(async (): Promise<Ad> => {
```

Comment:
Why don't we reuse or encapsulate ad fetching? I don't want it to happen in multiple places if possible

Reply:
Refactored it to we have one function for calling ads

Reply:
Also refactored the comment ad fetch, so it's in same utils file